### PR TITLE
Rename tapscript descriptors

### PR DESCRIPTION
**Pull request by jachiang**
As discussed in #123 

1) Tapscripts updated to use hash160 as a hashlock

2) Tapscripts updated to have the following descriptors:

* `pk`
* `pk_delay`
* `pk_hashlock`
* `pk_hashlock_delay`
* `csa`
* `csa_delay`
* `csa_hashlock`
* `csa_hashlock_delay`

3) Constructor methods have been updated accordingly...

* e.g. `TapLeaf().construct_csa_hashlock_delay(keys, hash, delay)`

### CODE REVIEW

**Comment by jnewbery** [comment]
Thanks for this, @jachiang ! The changes look mostly good, but you can help reviewers a lot by:

1. Using good, consistent commit logs (see the link I sent you here: https://github.com/bitcoin/bitcoin/pull/17288#pullrequestreview-308807117 for pointers)
2. don't break things between commits. For example, in this PR, you update the test_framework library in one commit and then update the notebooks in a subsequent commit. Between those commits, the notebooks wouldn't work. (This is especially important when a repo has automated tests because it would break git bisect)
3. don't mix up random clean-ups (eg spelling corrections, whitespace removal) in the same commit as functional changes.

Those things are slightly less important in this repo because there are only a couple of us working on it, we don't have automated tests, but it's best to get into good habits for when you're contributing to Bitcoin Core or other projects. It's also just good practice to make things as easy as possible for your reviewers.

File: 2.3-tapscript.ipynb [hunk 1]
```diff
@@ -396,9 +403,9 @@
...
    "metadata": {},
    "source": [
-    "#### 2.3.5 _Programming Exercise:_ Generate a 2-of-2 `csahasholder` tapscript\n",
+    "#### 2.3.5 _Programming Exercise:_ Generate a 2-of-2 `csa_hashlock_delay` tapscript\n",
     "\n",
-    "Construct a `csahasholder` tapscript with the following locking conditions:\n",
+    "Construct a `csa_hashlock_delay` tapscript with the following locking conditions:\n",
     "\n",
     "* 2-of-2 public keys\n",
     "* Hashlock with the preimage `sha256(b'secret')`\n",
```
Comment (jnewbery) [inline 2.3-tapscript.ipynb]: This is strange. You change the comment in the first commit and then revert it in a later commit.

Do you use the preimage `sha256(b'secret')` to enforce that the preimage provided in the witness is 32 bytes? If so, is that explained somewhere?

File: 2.3-tapscript.ipynb [hunk 2]
```diff
@@ -416,25 +423,26 @@
     "privkey2, pubkey2 = generate_key_pair()\n",
     "\n",
     "print(\"pubkey1: {}\".format(pubkey1.get_bytes().hex()))\n",
-    "print(\"pubkey2: {}\\n\".format(pubkey2.get_bytes().hex()))\n",
+    "print(\"pubkey2: {}\\n\".format(pubkey1.get_bytes().hex()))\n",
     "\n",
     "# Method: 32B preimage - sha256(bytes)\n",
-    "# Method: 20B digest - hashlib.new('ripemd160', bytes).digest()\n",
+    "# Method: 20B digest - hashlib.new('ripemd160', sha256(bytes)).digest()\n",
```
Comment (jnewbery) [inline 2.3-tapscript.ipynb]: Just use `hash160` from script.py and then remove the direct import of hashlib.

**Comment by jnewbery** [comment]
Can you restructure this PR as follows:

1. squash the commits _Changed tapscript hashlock: ripemd160 -> hash160._ and _Updated hash160 hashlock in chapter 2.3._ (with the review comments addressed)
2. Have one commit for renaming the descriptors. I think you can do this quite easily with a script:

```
sed -i -E -e 's/pkhasholder\b/pk_hashlock_delay/g' $(git grep -l pkhasholder)
sed -i -E -e 's/pkhash\b/pk_hashlock/g' $(git grep -l pkhash)
```
etc. Run the script and then paste it into the commit log so reviewers can verify the script.
3. Split the spelling/whitespace corrections into their own commit. This could potentially go into its own PR since it's not really related to the other changes in this PR.
4. Split the tapleaf constructor methods into their own commit. Again, this could go into its own PR.

**Comment by jachiang** [comment]
> Those things are slightly less important in this repo because there are only a couple of us working on it, we don't have automated tests, but it's best to get into good habits for when you're contributing to Bitcoin Core or other projects. It's also just good practice to make things as easy as possible for your reviewers.

@jnewbery Many thanks for the tips, much appreciated and will happily apply them here and in future PR's!

File: 2.3-tapscript.ipynb (hunk 1 above)
Comment (jachiang) [reply 2.3-tapscript.ipynb]: > Do you use the preimage sha256(b'secret') to enforce that the preimage provided in the witness is 32 bytes? If so, is that explained somewhere?

Yes there's a 32B size guard in the output script - I will add explainer that the hashlock does this to protect against unfeasibly large pre-images at spending time (e.g. atomic swap between two chains with different data push size limits, where correlating pre-image size is only valid on one-chain).

**Comment by jachiang** [comment]
@jnewbery Thanks for the review. I have applied most of the suggested fixes. Typos are split into #140 and this PR now build on top of that. It was easier to fix the typos first, since they affect the subsitution of tapscript descriptors.

Remaining TODO:
>  4. Split the tapleaf constructor methods into their own commit. Again, this could go into its own PR. 

Do you mean dedicating a PR to the renaming of the constructors to match the descriptor update? The tapscript-specific constructor methods should be renamed together with the descriptors, or I am missing something?

File: 2.3-tapscript.ipynb [hunk 3]
```diff
@@ -366,16 +386,22 @@
...
+    "\n",
+    "* `ts(csa_delay(k, [key0, key1, ...], hash, time))`\n",
+    "    * Witness: `[signature], [signature], ...`\n",
+    "    * Spendable after delay (with `nSequence > delay`)\n",
     "\n",
     "\n",
-    "* `ts(csahasholder(k, [key0, key1, ...], hash, time))`\n",
-    "    * Witness: `[signature], [signature], ..., [32B pre-image], [delay]`"
+    "* `ts(csa_hashlock_delay(k, [key0, key1, ...], hash, time))`\n",
+    "    * Witness: `[signature], [signature], ..., [32B pre-image], [delay]`\n",
```
Comment (jnewbery) [inline 2.3-tapscript.ipynb]: I don't think the witness includes `[delay]`

**Comment by jachiang** [comment]
Rebased after merging of #140.

File: 2.3-tapscript.ipynb [hunk 4]
```diff
@@ -6,14 +6,12 @@
...
    "source": [
-    "import hashlib\n",
-    "\n",
     "import util\n",
     "from test_framework.address import program_to_witness\n",
     "from test_framework.key import generate_key_pair\n",
     "from test_framework.messages import CTxInWitness, ser_string, sha256\n",
     "from test_framework.musig import generate_musig_key\n",
-    "from test_framework.script import TapLeaf, TapTree, TaprootSignatureHash, SIGHASH_ALL_TAPROOT"
+    "from test_framework.script import TapLeaf, TapTree, TaprootSignatureHash, SIGHASH_ALL_TAPROOT, hash160"
```
Comment (jnewbery) [inline 2.3-tapscript.ipynb]: nit: sort import names alphabetically

**Comment by jnewbery** [comment]
> > Split the tapleaf constructor methods into their own commit. Again, this could go into its own PR.

> Do you mean dedicating a PR to the renaming of the constructors to match the descriptor update? The tapscript-specific constructor methods should be renamed together with the descriptors, or I am missing something?

Sorry. I wasn't clear. Your original branch had changes like this, which constructed the tapleaf objects on a single line:

```
-    "tapscript_2a = TapLeaf()\n",
-    "tapscript_2b = TapLeaf()\n",
-    "tapscript_2c = TapLeaf()\n",
-    "tapscript_2d = TapLeaf()\n",
-    "tapscript_2e = TapLeaf()\n",
-    "tapscript_2f = TapLeaf()\n",
     "delay = 3*24*6\n",
-    "tapscript_2a.construct_csaolder(3, [main_pubkeyA, main_pubkeyB, backup_pubkeyD], delay)\n",
-    "tapscript_2b.construct_csaolder(3, [main_pubkeyA, main_pubkeyC, backup_pubkeyD], delay)\n",
-    "tapscript_2c.construct_csaolder(3, [main_pubkeyB, main_pubkeyC, backup_pubkeyD], delay)\n",
-    "tapscript_2d.construct_csaolder(3, [main_pubkeyA, main_pubkeyB, backup_pubkeyE], delay)\n",
-    "tapscript_2e.construct_csaolder(3, [main_pubkeyA, main_pubkeyC, backup_pubkeyE], delay)\n",
-    "tapscript_2f.construct_csaolder(3, [main_pubkeyB, main_pubkeyC, backup_pubkeyE], delay)\n",
+    "tapscript_2a = TapLeaf().construct_csa_delay(3, [main_pubkeyA, main_pubkeyB, backup_pubkeyD], delay)\n",
+    "tapscript_2b = TapLeaf().construct_csa_delay(3, [main_pubkeyA, main_pubkeyC, backup_pubkeyD], delay)\n",
+    "tapscript_2c = TapLeaf().construct_csa_delay(3, [main_pubkeyB, main_pubkeyC, backup_pubkeyD], delay)\n",
+    "tapscript_2d = TapLeaf().construct_csa_delay(3, [main_pubkeyA, main_pubkeyB, backup_pubkeyE], delay)\n",
+    "tapscript_2e = TapLeaf().construct_csa_delay(3, [main_pubkeyA, main_pubkeyC, backup_pubkeyE], delay)\n",
```

which I was trying to say should be in their own commit or PR.

**Comment by jnewbery** [comment]
sorry. Needs a rebase (at least the scripted name change should be easy!)

File: 2.3-tapscript.ipynb (hunk 3 above)
Comment (jachiang) [reply 2.3-tapscript.ipynb]: Fixed in #141.

**Comment by jachiang** [comment]
> Sorry. I wasn't clear. Your original branch had changes like this, which constructed the tapleaf objects on a single line:

**Comment by jachiang** [comment]
Oops. Didn't mean to close this. Spotty Internet.

**Comment by jachiang** [comment]
Ok, I have addressed nits, added the init/constructor chaining and rebased :)

**Comment by jnewbery** [comment]
Tested ACK ced902d5856661cdf6fa9ddfb34d05f1e86fe4f5

Great stuff. Thanks James!

**Pull request merged**